{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "pnyUCI1S4fGU"
   },
   "source": [
    "NGS files are all about big data. Fortunately, the bioinformatics community has developed tools to allow for the partial download of data. As part of the SAMtools/htslib package (http://www.htslib.org/),\n",
    "you can download tabix and bgzip, which will take care of data management. On the command line, perform the following operation:\n",
    "</br>\n",
    "command:\n",
    "</br>\n",
    "`apt-get install tabix`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "IrBoYxUy5dF-"
   },
   "source": [
    "The bellow line will partially download the VCF file for chromosome 22 (up to 17 Mbp) of the 1,000 Genomes Project. Then, bgzip will compress it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Channels:\n",
      " - conda-forge\n",
      " - bioconda\n",
      "Platform: linux-64\n",
      "Collecting package metadata (repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "    current version: 24.5.0\n",
      "    latest version: 25.11.0\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c conda-forge conda\n",
      "\n",
      "\n",
      "\n",
      "# All requested packages already installed.\n",
      "\n",
      "\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "conda install -c conda-forge -c bioconda tabix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2025-12-01 06:50:19--  https://github.com/samtools/htslib/releases/download/1.22.1/htslib-1.22.1.tar.bz2\n",
      "Resolving github.com (github.com)... 140.82.112.3\n",
      "Connecting to github.com (github.com)|140.82.112.3|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 Found\n",
      "Location: https://release-assets.githubusercontent.com/github-production-release-asset/4339773/2fcb9687-d889-4790-a4bd-efb525669bbd?sp=r&sv=2018-11-09&sr=b&spr=https&se=2025-12-01T07%3A47%3A23Z&rscd=attachment%3B+filename%3Dhtslib-1.22.1.tar.bz2&rsct=application%2Foctet-stream&skoid=96c2d410-5711-43a1-aedd-ab1947aa7ab0&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skt=2025-12-01T06%3A47%3A15Z&ske=2025-12-01T07%3A47%3A23Z&sks=b&skv=2018-11-09&sig=xGazf%2FAYFnQx4xPuU4p8QlFjVAOgxIyk0TJ4DYEldzw%3D&jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmVsZWFzZS1hc3NldHMuZ2l0aHVidXNlcmNvbnRlbnQuY29tIiwia2V5Ijoia2V5MSIsImV4cCI6MTc2NDU3MjEyMCwibmJmIjoxNzY0NTcxODIwLCJwYXRoIjoicmVsZWFzZWFzc2V0cHJvZHVjdGlvbi5ibG9iLmNvcmUud2luZG93cy5uZXQifQ.UlAvZzWhTthd35slO_aXI4D9qKBqLxjzNzUDwlZZNHE&response-content-disposition=attachment%3B%20filename%3Dhtslib-1.22.1.tar.bz2&response-content-type=application%2Foctet-stream [following]\n",
      "--2025-12-01 06:50:20--  https://release-assets.githubusercontent.com/github-production-release-asset/4339773/2fcb9687-d889-4790-a4bd-efb525669bbd?sp=r&sv=2018-11-09&sr=b&spr=https&se=2025-12-01T07%3A47%3A23Z&rscd=attachment%3B+filename%3Dhtslib-1.22.1.tar.bz2&rsct=application%2Foctet-stream&skoid=96c2d410-5711-43a1-aedd-ab1947aa7ab0&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skt=2025-12-01T06%3A47%3A15Z&ske=2025-12-01T07%3A47%3A23Z&sks=b&skv=2018-11-09&sig=xGazf%2FAYFnQx4xPuU4p8QlFjVAOgxIyk0TJ4DYEldzw%3D&jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmVsZWFzZS1hc3NldHMuZ2l0aHVidXNlcmNvbnRlbnQuY29tIiwia2V5Ijoia2V5MSIsImV4cCI6MTc2NDU3MjEyMCwibmJmIjoxNzY0NTcxODIwLCJwYXRoIjoicmVsZWFzZWFzc2V0cHJvZHVjdGlvbi5ibG9iLmNvcmUud2luZG93cy5uZXQifQ.UlAvZzWhTthd35slO_aXI4D9qKBqLxjzNzUDwlZZNHE&response-content-disposition=attachment%3B%20filename%3Dhtslib-1.22.1.tar.bz2&response-content-type=application%2Foctet-stream\n",
      "Resolving release-assets.githubusercontent.com (release-assets.githubusercontent.com)... 185.199.110.133, 185.199.109.133, 185.199.108.133, ...\n",
      "Connecting to release-assets.githubusercontent.com (release-assets.githubusercontent.com)|185.199.110.133|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 4840271 (4.6M) [application/octet-stream]\n",
      "Saving to: ‘htslib-1.22.1.tar.bz2’\n",
      "\n",
      "htslib-1.22.1.tar.b 100%[===================>]   4.62M  17.4MB/s    in 0.3s    \n",
      "\n",
      "2025-12-01 06:50:20 (17.4 MB/s) - ‘htslib-1.22.1.tar.bz2’ saved [4840271/4840271]\n",
      "\n",
      "tar: htslib-1.19.tar.bz2: Cannot open: No such file or directory\n",
      "tar: Error is not recoverable: exiting now\n",
      "[Errno 2] No such file or directory: 'htslib-1.19'\n",
      "/jetstream2/scratch/main/jobs/72985031/working/jupyter/htslib-1.19\n",
      "/bin/bash: line 1: autoreconf: command not found\n",
      "checking for gcc... gcc\n",
      "checking whether the C compiler works... yes\n",
      "checking for C compiler default output file name... a.out\n",
      "checking for suffix of executables... \n",
      "checking whether we are cross compiling... no\n",
      "checking for suffix of object files... o\n",
      "checking whether the compiler supports GNU C... yes\n",
      "checking whether gcc accepts -g... yes\n",
      "checking for gcc option to enable C11 features... none needed\n",
      "checking for ranlib... ranlib\n",
      "checking for grep that handles long lines and -e... /usr/bin/grep\n",
      "checking for C compiler warning flags... -Wall\n",
      "checking for gcc options needed to detect all undeclared functions... none needed\n",
      "checking for stdio.h... yes\n",
      "checking for stdlib.h... yes\n",
      "checking for string.h... yes\n",
      "checking for inttypes.h... yes\n",
      "checking for stdint.h... yes\n",
      "checking for strings.h... yes\n",
      "checking for sys/stat.h... yes\n",
      "checking for sys/types.h... yes\n",
      "checking for unistd.h... yes\n",
      "checking for sys/param.h... yes\n",
      "checking whether _XOPEN_SOURCE is declared... no\n",
      "checking C compiler flags needed for sse4.1... -msse4.1 -mssse3 -mpopcnt\n",
      "checking C compiler flags needed for avx2... -mavx2 -mpopcnt\n",
      "checking C compiler flags needed for avx512f... -mavx512f -mpopcnt\n",
      "checking for pkg-config... no\n",
      "checking for gcc option to enable large file support... none needed\n",
      "checking shared library type for unknown-Linux... plain .so\n",
      "checking whether the linker supports versioned symbols... yes\n",
      "checking whether the compiler accepts -fvisibility=hidden... yes\n",
      "checking build system type... x86_64-pc-linux-gnu\n",
      "checking host system type... x86_64-pc-linux-gnu\n",
      "checking for getpagesize... yes\n",
      "checking for working mmap... yes\n",
      "checking for gmtime_r... yes\n",
      "checking for fsync... yes\n",
      "checking for drand48... yes\n",
      "checking for srand48_deterministic... no\n",
      "checking whether fdatasync is declared... yes\n",
      "checking for fdatasync... yes\n",
      "checking for library containing log... -lm\n",
      "checking for zlib.h... no\n",
      "checking for inflate in -lz... no\n",
      "configure: error: zlib development files not found\n",
      "\n",
      "HTSlib uses compression routines from the zlib library <http://zlib.net>.\n",
      "Building HTSlib requires zlib development files to be installed on the build\n",
      "machine; you may need to ensure a package such as zlib1g-dev (on Debian or\n",
      "Ubuntu Linux) or zlib-devel (on RPM-based Linux distributions or Cygwin)\n",
      "is installed.\n",
      "\n",
      "FAILED.  This error must be resolved in order to build HTSlib successfully.\n",
      "/bin/bash: line 1: make: command not found\n",
      "/bin/bash: line 1: make: command not found\n"
     ]
    }
   ],
   "source": [
    "# !wget https://github.com/samtools/htslib/releases/download/1.22.1/htslib-1.22.1.tar.bz2\n",
    "# !tar -xf htslib-1.19.tar.bz2\n",
    "# %cd htslib-1.19\n",
    "# !autoreconf -i\n",
    "# !./configure --enable-libcurl\n",
    "# !make\n",
    "# !make install"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "id": "6t7gbA0w2rwL"
   },
   "outputs": [],
   "source": [
    "!tabix -fh https://ftp.ncbi.nlm.nih.gov/1000genomes/ftp/release/20130502/supporting/vcf_with_sample_level_annotation/ALL.chr22.phase3_shapeit2_mvncall_integrated_v5_extra_anno.20130502.genotypes.vcf.gz 22:1-17000000|bgzip -c > genotypes.vcf.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "2zNwT61B5k-P"
   },
   "source": [
    "Then, We should create an index, which we will need for direct access to a section of the genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "id": "LQzb8YJQ3iF8"
   },
   "outputs": [],
   "source": [
    "!tabix -p vcf genotypes.vcf.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "mVLUB5Nv6RjI"
   },
   "source": [
    "**Let’s start by inspecting the information that we can get per record:**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "hvdEta9t7-7I"
   },
   "source": [
    "We start by inspecting the annotations that are available for each record (remember that each record encodes a variant, such as SNP, CNV, INDELs, and so on, and the state of that variant per sample). \n",
    "</br>\n",
    "At the variant (record) level, we find AC—the total number of ALT alleles in called genotypes, AF—the estimated allele frequency, NS—the number of samples with data, AN—the total number of alleles in called genotypes, and DP—the total read depth. \n",
    "</br>\n",
    "There are others,\n",
    "but they are mostly specific to the 1,000 Genomes Project (here, we will try to be as general as\n",
    "possible). Your own dataset may have more annotations (or none of these).\n",
    "At the sample level, there are only two annotations in this file: GT—genotype, and DP—the\n",
    "per-sample read depth. You have the per-variant (total) read depth and the per-sample read\n",
    "depth; be sure not to confuse both."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Channels:\n",
      " - conda-forge\n",
      " - bioconda\n",
      "Platform: linux-64\n",
      "Collecting package metadata (repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "    current version: 24.5.0\n",
      "    latest version: 25.11.0\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c conda-forge conda\n",
      "\n",
      "\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /opt/conda\n",
      "\n",
      "  added / updated specs:\n",
      "    - cyvcf2\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    click-8.3.1                |     pyh8f84b5b_1          95 KB  conda-forge\n",
      "    coloredlogs-15.0.1         |     pyhd8ed1ab_4          43 KB  conda-forge\n",
      "    cyvcf2-0.31.4              |  py311h6e87821_0         998 KB  bioconda\n",
      "    htslib-1.22.1              |       h566b1c6_0         3.1 MB  bioconda\n",
      "    humanfriendly-10.0         |     pyh707e725_8          72 KB  conda-forge\n",
      "    libcurl-8.14.1             |       h332b0f4_0         439 KB  conda-forge\n",
      "    liblzma-5.8.1              |       hb9d3cd8_2         110 KB  conda-forge\n",
      "    liblzma-devel-5.8.1        |       hb9d3cd8_2         430 KB  conda-forge\n",
      "    libnghttp2-1.64.0          |       h161d5f1_0         632 KB  conda-forge\n",
      "    libssh2-1.11.1             |       hcf80075_0         298 KB  conda-forge\n",
      "    numpy-2.3.5                |  py311h2e04523_0         9.0 MB  conda-forge\n",
      "    scipy-1.16.0               |  py311h2d3ef60_0        16.1 MB  conda-forge\n",
      "    xz-5.8.1                   |       hbcc6ac9_2          23 KB  conda-forge\n",
      "    xz-gpl-tools-5.8.1         |       hbcc6ac9_2          33 KB  conda-forge\n",
      "    xz-tools-5.8.1             |       hb9d3cd8_2          94 KB  conda-forge\n",
      "    zstandard-0.25.0           |  py311haee01d2_1         456 KB  conda-forge\n",
      "    zstd-1.5.7                 |       h3691f8a_5         537 KB  conda-forge\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:        32.4 MB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  click              conda-forge/noarch::click-8.3.1-pyh8f84b5b_1 \n",
      "  coloredlogs        conda-forge/noarch::coloredlogs-15.0.1-pyhd8ed1ab_4 \n",
      "  cyvcf2             bioconda/linux-64::cyvcf2-0.31.4-py311h6e87821_0 \n",
      "  humanfriendly      conda-forge/noarch::humanfriendly-10.0-pyh707e725_8 \n",
      "  liblzma            conda-forge/linux-64::liblzma-5.8.1-hb9d3cd8_2 \n",
      "  liblzma-devel      conda-forge/linux-64::liblzma-devel-5.8.1-hb9d3cd8_2 \n",
      "  xz-gpl-tools       conda-forge/linux-64::xz-gpl-tools-5.8.1-hbcc6ac9_2 \n",
      "  xz-tools           conda-forge/linux-64::xz-tools-5.8.1-hb9d3cd8_2 \n",
      "\n",
      "The following packages will be UPDATED:\n",
      "\n",
      "  htslib                                    1.21-h5efdd21_0 --> 1.22.1-h566b1c6_0 \n",
      "  libcurl                                  8.9.0-hdb1bdb2_0 --> 8.14.1-h332b0f4_0 \n",
      "  libnghttp2                              1.58.0-h47da74e_1 --> 1.64.0-h161d5f1_0 \n",
      "  libssh2                                 1.11.0-h0841786_0 --> 1.11.1-hcf80075_0 \n",
      "  numpy                               2.1.1-py311h71ddf71_0 --> 2.3.5-py311h2e04523_0 \n",
      "  scipy                              1.14.1-py311he1f765f_0 --> 1.16.0-py311h2d3ef60_0 \n",
      "  xz                                       5.2.6-h166bdaf_0 --> 5.8.1-hbcc6ac9_2 \n",
      "  zstandard                          0.23.0-py311h5cd10c7_0 --> 0.25.0-py311haee01d2_1 \n",
      "  zstd                                     1.5.6-ha6fb4c9_0 --> 1.5.7-h3691f8a_5 \n",
      "\n",
      "\n",
      "\n",
      "Downloading and Extracting Packages:\n",
      "scipy-1.16.0         | 16.1 MB   |                                       |   0% \n",
      "numpy-2.3.5          | 9.0 MB    |                                       |   0% \u001b[A\n",
      "\n",
      "htslib-1.22.1        | 3.1 MB    |                                       |   0% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "cyvcf2-0.31.4        | 998 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "libnghttp2-1.64.0    | 632 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "zstd-1.5.7           | 537 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "zstandard-0.25.0     | 456 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libcurl-8.14.1       | 439 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "liblzma-devel-5.8.1  | 430 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libssh2-1.11.1       | 298 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "liblzma-5.8.1        | 110 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "click-8.3.1          | 95 KB     |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "xz-tools-5.8.1       | 94 KB     |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "humanfriendly-10.0   | 72 KB     |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "coloredlogs-15.0.1   | 43 KB     |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "xz-gpl-tools-5.8.1   | 33 KB     |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "xz-5.8.1             | 23 KB     |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "scipy-1.16.0         | 16.1 MB   |                                       |   0% \u001b[A\n",
      "\n",
      "htslib-1.22.1        | 3.1 MB    | #############9                        |  38% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "cyvcf2-0.31.4        | 998 KB    | #######1                              |  19% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "zstd-1.5.7           | 537 KB    | #1                                    |   3% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "zstandard-0.25.0     | 456 KB    | #2                                    |   4% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "numpy-2.3.5          | 9.0 MB    | ############                          |  33% \u001b[A\n",
      "\n",
      "scipy-1.16.0         | 16.1 MB   | ###9                                  |  11% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libcurl-8.14.1       | 439 KB    | #3                                    |   4% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "liblzma-devel-5.8.1  | 430 KB    | #3                                    |   4% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "scipy-1.16.0         | 16.1 MB   | #######7                              |  21% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libssh2-1.11.1       | 298 KB    | #9                                    |   5% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "liblzma-5.8.1        | 110 KB    | #####3                                |  15% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "zstandard-0.25.0     | 456 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "zstandard-0.25.0     | 456 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "libnghttp2-1.64.0    | 632 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "libnghttp2-1.64.0    | 632 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "click-8.3.1          | 95 KB     | ######2                               |  17% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "xz-tools-5.8.1       | 94 KB     | ######2                               |  17% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "scipy-1.16.0         | 16.1 MB   | ###########3                          |  31% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "coloredlogs-15.0.1   | 43 KB     | #############8                        |  37% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "xz-gpl-tools-5.8.1   | 33 KB     | #################8                    |  48% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "xz-5.8.1             | 23 KB     | #########################2            |  68% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "cyvcf2-0.31.4        | 998 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\n",
      "numpy-2.3.5          | 9.0 MB    | ##############################5       |  83% \u001b[A\n",
      "\n",
      "\n",
      "scipy-1.16.0         | 16.1 MB   | ################5                     |  45% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libcurl-8.14.1       | 439 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libcurl-8.14.1       | 439 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "zstd-1.5.7           | 537 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "scipy-1.16.0         | 16.1 MB   | ###############################7      |  86% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "liblzma-5.8.1        | 110 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "liblzma-5.8.1        | 110 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "liblzma-devel-5.8.1  | 430 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "liblzma-devel-5.8.1  | 430 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "click-8.3.1          | 95 KB     | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "click-8.3.1          | 95 KB     | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "xz-tools-5.8.1       | 94 KB     | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "xz-tools-5.8.1       | 94 KB     | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libssh2-1.11.1       | 298 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "libssh2-1.11.1       | 298 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "xz-5.8.1             | 23 KB     | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "xz-5.8.1             | 23 KB     | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "humanfriendly-10.0   | 72 KB     | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "humanfriendly-10.0   | 72 KB     | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "coloredlogs-15.0.1   | 43 KB     | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "coloredlogs-15.0.1   | 43 KB     | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "xz-gpl-tools-5.8.1   | 33 KB     | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "xz-gpl-tools-5.8.1   | 33 KB     | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "htslib-1.22.1        | 3.1 MB    | ##################################### | 100% \u001b[A\u001b[A\n",
      "                                                                                \u001b[A\n",
      "                                                                                \u001b[A\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n",
      "\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "conda install cyvcf2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "7Q5i5gm85-Dt",
    "outputId": "788e6862-8fdb-48ac-890c-440c21e15588"
   },
   "outputs": [
    {
     "ename": "OSError",
     "evalue": "genotypes.vcf.gz is not valid bcf or vcf (format: 15 mode: r)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mOSError\u001b[0m                                   Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# required package\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mcyvcf2\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m VCF\n\u001b[0;32m----> 3\u001b[0m v \u001b[38;5;241m=\u001b[39m \u001b[43mVCF\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mgenotypes.vcf.gz\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      4\u001b[0m rec \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mnext\u001b[39m(v)\n\u001b[1;32m      5\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mVariant Level information\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[0;32m/opt/conda/lib/python3.11/site-packages/cyvcf2/cyvcf2.pyx:278\u001b[0m, in \u001b[0;36mcyvcf2.cyvcf2.VCF.__init__\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m/opt/conda/lib/python3.11/site-packages/cyvcf2/cyvcf2.pyx:211\u001b[0m, in \u001b[0;36mcyvcf2.cyvcf2.HTSFile._open_htsfile\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mOSError\u001b[0m: genotypes.vcf.gz is not valid bcf or vcf (format: 15 mode: r)"
     ]
    }
   ],
   "source": [
    "# required package\n",
    "from cyvcf2 import VCF\n",
    "v = VCF('genotypes.vcf.gz')\n",
    "rec = next(v)\n",
    "print('Variant Level information')\n",
    "info = rec.INFO\n",
    "for info in rec.INFO:\n",
    "  print(info)\n",
    "print('Sample Level information')\n",
    "for fmt in rec.FORMAT:\n",
    "  print(fmt)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "hM46pI8m7hXg"
   },
   "source": [
    "**Now that we know what information is available, let’s inspect a single VCF record:**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "OxOxi21576ah"
   },
   "source": [
    "We will start by retrieving the standard information: the chromosome, position, ID, reference base\n",
    "(typically just one) and alternative bases (you can have more than one, but it’s not uncommon\n",
    "as a first filtering approach to only accept a single ALT, for example, only accept biallelic\n",
    "SNPs), quality (as you might expect, Phred-scaled), and filter status. Regarding the filter status,\n",
    "remember that whatever the VCF file says, you may still want to apply extra filters (as in the\n",
    "next recipe, Studying genome accessibility and filtering SNP data).\n",
    "We then print the additional variant-level information (AC, AS, AF, AN, DP, and so on),\n",
    "followed by the sample format (in this case, DP and GT). Finally, we count the number of\n",
    "samples and inspect a single sample to check whether it was called for this variant. Also, the\n",
    "reported alleles, heterozygosity, and phasing status (this dataset happens to be phased, which\n",
    "is not that common) are included."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "aS1xqzny7k4S",
    "outputId": "46cf645a-91db-47ee-952d-36c7760e07c8"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2504\n",
      "22 16050075 None A ['G'] 100.0 None\n",
      "<cyvcf2.cyvcf2.INFO object at 0x7f86b9ddff90>\n",
      "['GT', 'DP']\n",
      "True\n",
      "A|A [0, 0] True\n",
      "1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse_info] INFO 'SAS_AF' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse_info] INFO 'EAS_AF' is not defined in the header, assuming Type=String\n"
     ]
    }
   ],
   "source": [
    "v = VCF('genotypes.vcf.gz')\n",
    "samples = v.samples\n",
    "print(len(samples))\n",
    "variant = next(v)\n",
    "print(variant.CHROM, variant.POS, variant.ID, variant.\n",
    "REF, variant.ALT, variant.QUAL, variant.FILTER)\n",
    "print(variant.INFO)\n",
    "print(variant.FORMAT)\n",
    "print(variant.is_snp)\n",
    "str_alleles = variant.gt_bases[0]\n",
    "alleles = variant.genotypes[0][0:2]\n",
    "is_phased = variant.genotypes[0][2]\n",
    "print(str_alleles, alleles, is_phased)\n",
    "print(variant.format('DP')[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "eS5_QUyo7zOI"
   },
   "source": [
    "**Let’s check the type of variant and the number of nonbiallelic SNPs in a single pass:**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "OdND2Y-d8JKV"
   },
   "source": [
    "We will now use the now-common Python default dictionary. We find that this dataset has\n",
    "INDELs, CNVs, and—of course—SNPs (roughly two-thirds being transitions with one-third\n",
    "transversions). There is a residual number (79) of triallelic SNPs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "cZsOLSSI8U0o",
    "outputId": "1f63cd28-f33d-4b8d-b330-32b73c9bcc44"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse_info] INFO 'SAS_AF' is not defined in the header, assuming Type=String\n",
      "[W::vcf_parse_info] INFO 'EAS_AF' is not defined in the header, assuming Type=String\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "defaultdict(<class 'int'>, {('snp', 'ts'): 10054, ('snp', 'tv'): 5917, ('sv', 'CNV'): 2, ('indel', 'del'): 273, ('snp', 'unknown'): 79, ('indel', 'ins'): 127, ('indel', 'unknown'): 13, ('sv', 'DEL'): 6, ('sv', 'SVA'): 1})\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "f = VCF('genotypes.vcf.gz')\n",
    "my_type = defaultdict(int)\n",
    "num_alts = defaultdict(int)\n",
    "for variant in f:\n",
    "  my_type[variant.var_type, variant.var_subtype] += 1\n",
    "  if variant.var_type == 'snp':\n",
    "    num_alts[len(variant.ALT)] += 1\n",
    "print(my_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
